Add documentation on releasing new data (#1081)
PublicMapping/districtbuilder: deployment/README.md

### Variables

| Environment | Cluster                | Cluster VPC             | Security Group         |
| ----------- | ---------------------- | ----------------------- | ---------------------- |
| Staging     | `ecsStagingCluster`    | `vpc-04d3fda63dfc36e58` | `sg-0eb1ba122ccc68ca7` |
| Production  | `ecsProductionCluster` | `vpc-039833dc732e496a1` | `sg-05a0cca2a9f5b57a3` |

> ## Release New State Data
> 
> To release new data, you will need a database administration desktop application for PostgreSQL like PG Admin. You can find information on configuring PG Admin for DistrictBuilder's databases [here](https://github.com/PublicMapping/districtbuilder/blob/develop/deployment/configure_pgadmin.md).
> 
> You will need to update both the staging and production databases, first staging and then production, in order to run quality assurance on the data before posting it to the production site.
> 
> ### The Data
> 
> The list of states with new data can be found [here](https://docs.google.com/spreadsheets/d/1khiNAQp4UnfkgfvRSbfjDU9CU2zJrJ0pMRUeF5a4cuM/edit#gid=645349603) highlighted in yellow. Before applying these changes to the staging database, you will need to make test maps of the affected states. Open [staging](https://app.staging.districtbuilder.org/) and make a new map of each state highlighted in yellow, adding both counties and blockgroups to a few districts.

Reviewer: A few suggested changes:
 - Added a preamble to explain why we're doing this
 - state -> region
 - removed references to the google sheet. Not all regions are VEST data / coordinated using the Google sheet, just the main 50 states - we also have county & municipal regions, as well as at least 1 alternative state region

```suggestion
## Updating Region Data

Careful steps need to be taken when adding or updating regions to prevent site downtime, and additionally when updating existing regions to avoid unintentionally storing stale data or breaking existing maps.

### Pre-update steps

Before applying changes to the staging database or S3 assets, you will need to make test maps of the affected regions. Open [staging](https://app.staging.districtbuilder.org/) and make a new map of each region being updated, adding both counties and blockgroups to a few districts.

### Database access
To release new data, you will need to access the staging & production databases. You may prefer to use a database administration desktop application for PostgreSQL like PG Admin. You can find information on configuring PG Admin for DistrictBuilder's databases [here](https://github.com/PublicMapping/districtbuilder/blob/develop/deployment/configure_pgadmin.md).

You will need to update both the staging and production databases, first staging and then production, in order to run quality assurance on the data before posting it to the production site.
```

> ### The Method
> 
> There are two ways to update the data in staging: the best practice way and the fast way. While staging can be done either way depending on the given timeframe, updating the data in production should _always_ happen the best practice way. If you do not follow the best practice way for production data, the site will crash.
> 
> **The Best Practice Way**
> 
> The best practice way is to overwrite the data on S3 and force reload once all data has been overwritten. To do so, for each state with updated data, do the following:
> 
> - Make a backup of the data by running `aws s3 –profile=district-builder cp -r <existing URI> <existing URI>-backup-<yyyy-mm-dd>`
> - Copy the new data over the old with `aws s3 –profile=district-builder cp -r <new URI> <existing URI>`
> 
> **The Fast Way**
> 
> The faster but less ideal way for use only in staging is to just update the URI in the `s3_uri` column of the database in PG Admin. Then save the database.
> 
> Once you have updated the data, follow the instructions in the [Restart Services](https://github.com/PublicMapping/districtbuilder/tree/develop/deployment#restart-services) section below.

Reviewer: Mostly text tweaks here, expanded on the warning up top:

```suggestion
### Update process

For existing regions there are two ways to update the region to use new data: by using S3 to overwrite the existing data or by updating the database.

Note that updates to the `s3_uri` column on production for an existing region **will** cause API calls for that region as well as healthchecks to fail until data is loaded, potentially resulting in a production outage. For this reason, it is recommended to _only_ ever make updates to the `region_config.s3_uri` column on the staging database.

#### Overwriting data on S3

For existing regions, when updating data the best practice way is to overwrite the data on S3 and reload tasks once all data has been overwritten. To do so, for each region with updated data, do the following:

- Make a backup of the data by running `aws s3 –profile=district-builder cp -r <existing URI> <existing URI>-backup-<yyyy-mm-dd>`
- Copy the new data over the old with `aws s3 –profile=district-builder cp -r <new URI> <existing URI>`

#### Updating the database

The faster but less ideal way for use only in staging is to just update the URI in the `s3_uri` column of the database in PG Admin. Then save the database.

#### Reloading services

Once you have updated the data, follow the instructions in the [Restart Services](https://github.com/PublicMapping/districtbuilder/tree/develop/deployment#restart-services) section below.
```

### Quality Assurance

> Once you have restarted the services for staging, you will need to return to the test maps you made. Open each to check that it loads and then make an edit or two, both at a block group and county level. If nothing breaks, that state's data is good to go up on production.

Reviewer: ```suggestion
Once you have restarted the services for staging, you will need to return to the test maps you made. Open each to check that it loads and then make an edit or two, both at a block group and county level. If nothing breaks, that region's data is good to go up on production.
```

### Finishing New Data Release

> Once you have completed the data update for both staging and production, copy each updated state's URI listed in the production database to the `production s3 url` column in the Google Sheet to mark it as complete.

Reviewer: ```suggestion
Once you have completed the data update for both staging and production, copy each updated region's URI listed in the production database to the `production s3 url` column in the Google Sheet to mark it as complete.
```

It will take some time (approx. 20-25 mins) for the new tasks to finish loading all TopoJSON data. You can monitor the progress by looking for `HealthCheckService` logs in the [`logStagingApp` / `logProductionApp` Cloudwatch logs](https://console.aws.amazon.com/cloudwatch/home?region=us-east-1#logsV2:log-groups) or by looking at the number of healthy hosts in the [appropriate target group](https://console.aws.amazon.com/ec2/v2/home?region=us-east-1#TargetGroups:).

> If while monitoring the logs, you see _one_ error about topology failing to load, wait and it will try again (health check failures during this process are normal). If you see _multiple_ errors about topology failing to load, check both that state's URI and the contents of the file. Once a fix is made, you will need to kill the task following instructions from the Stopping Tasks section below and let it restart itself.

Reviewer: ```suggestion
If while monitoring the logs, you see _one_ error about topology failing to load, wait and it will try again (health check failures during this process are normal). If you see _multiple_ errors about topology failing to load, check both that region's `s3_uri` in the database as well the contents of the file. Once a fix is made, you will need to kill the task following instructions from the Stopping Tasks section below and let it restart itself.
```
